_moderngl.py

```diff
@@ -366,5 +366,168 @@ def make_varying(name, number, array_length, dimension):
+            pointer_allowed_types[spv[idx + 1]] = spv[idx + 3]
+            # --------------------^pointer_id^^----^^^type_id^^==
+            #print(spv[idx + 1], spv[idx + 2], spv[idx + 3])
+        
+        if opcode == 19:  # OpTypeVoid
+            allowed_types[spv[idx + 1]] = ['void', -1]
+        if opcode == 20:  # OpTypeBool
+            allowed_types[spv[idx + 1]] = ['bool', -1]
+         
+        if opcode == 21:  # OpTypeInt
+            addware_unsigned = ''
+            if spv[idx + 3] == 1:
+                 addware_unsigned = 'u'
+            allowed_types[spv[idx + 1]] = [f'{addware_unsigned}i{spv[idx + 2]//8}', -1]
+            
+        if opcode == 22:  # OpTypeFloat
+            allowed_types[spv[idx + 1]] = [f'f{spv[idx + 2]//8}', -1]
+            
+        if opcode == 23:  # OpTypeVector
+            allowed_types[spv[idx + 1]] = [f'vec{spv[idx + 3]}', spv[idx + 2]]
+            
+        if opcode == 24:  # OpTypeMatrix
+            allowed_types[spv[idx + 1]] = [f'mat{spv[idx + 3]}', spv[idx + 2]]
+        
+        idx += args
+    ##################
+
+    # == Assembly types == #
+    # Some types of variables have pointers to other types of variables used in them.
+    # So we use endless research to identify complete types of variables.
+    def assembly(type_id):
+        typ, thrw_typ = allowed_types[type_id]
+        if thrw_typ != -1:
+            add_typ = assembly(thrw_typ)
+            allowed_types[type_id][0] = add_typ+' '+typ
+            allowed_types[type_id][1] = -1
+        return typ
+
+    for type_id, config in allowed_types.items():
+        assembly(type_id)
+
+    for ids, pointer_type_id in pointer_variable.items():
+        try:
+            extracted_types[ids] = allowed_types[pointer_allowed_types[pointer_type_id]][0]
+        except KeyError:
+            pass
+    ###############
+
+    # == Making a whole list == #
+    exrtacted_general_ids = [] # id, id, id ...
+    exrtacted_general_ids = list( # get all identifiers for variables
+        set(extracted_location.keys()) | set(extracted_types.keys()) | \
+        set(extracted_storage_class_id.keys()) | set(extracted_names.keys()))
+    # exrtacted_general_ids.sort()  # if needed for sorting locations
+
+    extracted_collected = {}
+    for ids in exrtacted_general_ids:
+        to_add = {'name':'', 'class':-1, 'type':'', 'location':-1}
+        if ids in extracted_names:
+            to_add['name'] = extracted_names[ids]
+        if ids in extracted_storage_class_id:
+            to_add['class'] = extracted_storage_class_id[ids]
+        if ids in extracted_types:
+            to_add['type'] = extracted_types[ids]
+        if ids in extracted_location:
+            to_add['location'] = extracted_location[ids]
+        extracted_collected[ids]=to_add
+    ##################
+
+    # == Conversion to the moderngl type == #
+    mgl_attr_table = {  # py_type : mgl_type
```

Comment:
move this out of the function just like `ATTRIBUTE_LOOKUP_TABLE`

Reply:
OK